### PR TITLE
Use the proper varargs definition of libcurl functions

### LINKED ISSUE #180: Incorrect function definitions
The function definitions for `curl_easy_getinfo`, `curl_easy_setopt`, `curl_multi_setopt` in [`lib/ethon/curls/functions.rb`](https://github.com/typhoeus/ethon/blob/b382aad1083755cc8e2b40886d6b4bc6d77fcb8c/lib/ethon/curls/functions.rb) are incorrect.

They are declared in curl as follows.

```c
CURL_EXTERN CURLcode curl_easy_getinfo(CURL *curl, CURLINFO info, ...);
CURL_EXTERN CURLcode curl_easy_setopt(CURL *curl, CURLoption option, ...);
CURL_EXTERN CURLMcode curl_multi_setopt(CURLM *multi_handle,
                                        CURLMoption option, ...);
```

Meaning that they should be declared something like:

```ruby
attach_function :easy_getinfo, :curl_easy_getinfo, [:pointer, :info, :varargs], :easy_code
attach_function :easy_setopt, :curl_easy_setopt, [:pointer, :easy_option, :varargs], :easy_code
attach_function :multi_setopt, :curl_multi_setopt, [:pointer, :multi_option, :varargs], :multi_code
```

On some platforms, `curl_easy_getinfo` being declared as `[:pointer, :info, :varargs], :easy_code`, or (as it is currently in ethon) `[:pointer, :info, :pointer], :easy_code`, won't change how the method is called, but on some platforms (like on Apple Silicon / Apple M1), the ABI is different:

https://developer.apple.com/documentation/apple_silicon/addressing_architectural_differences_in_your_macos_code#3616882

> The `x86_64` and `arm64` architectures have different calling conventions for variadic functions—functions with a variable number of parameters. On `x86_64`, the compiler treats fixed and variadic parameters the same, placing parameters in registers first and only using the stack when no more registers are available. On `arm64`, the compiler always places variadic parameters on the stack, regardless of whether registers are available. If you implement a function with fixed parameters, but redeclare it with variadic parameters, the mismatch causes unexpected behavior at runtime.

Fixing it on the ethon side is not that hard (quick and dirty patch [here](https://gist.github.com/vincentisambart/c78e9df22d0d4851c1e1d150ac26e5d1)), but unfortunately that is not enough: it seems the ffi gem currently doesn't support passing callbacks via varargs (ffi/ffi#732).

So the ffi gem has to be fixed first.

### PR DESCRIPTION
**Pull request by vincentisambart**
closes #180

As explained in #180, the definition of some libcurl functions in `lib/ethon/curls/functions.rb` was not in line with the definition in the libcurl C headers. This PR fixes this problem.

As I was told in https://github.com/typhoeus/ethon/issues/180#issuecomment-788652324 that supporting older versions of ffi was not needed, so I tried to go for what seemed to end with the cleanest code, requiring ffi 1.5.0.

In the [patch](https://gist.github.com/vincentisambart/c78e9df22d0d4851c1e1d150ac26e5d1) I initially wrote and linked in the issue, I went for the safest route, only modifying `lib/ethon/curls/functions.rb`, adding `easy_setopt_*` and `multi_setopt_*` functions with the old signatures that called the new varargs version. This PR modifies more code and so is probably more risky, but I think in the long run it ends up with code more easy to follow. But if you want I can revert back to the safer change.

Note that as I'm writing this, it seems preparation for the release has started on the ffi repository, but ffi 1.5.0 has not been released yet. I'll remove the draft status of this PR as soon as ffi 1.5.0 is released.

### CODE REVIEW

File: ethon.gemspec
Comment (vincentisambart) [inline ethon.gemspec:21]: Passing callbacks via varargs requires ffi 1.5.0

**Comment by vincentisambart** [comment]
The CI failure is due to ffi 1.5.0 not being used. I also need to do a `bundle update` but that requires ffi 1.5.0 being released first.

**Comment by vincentisambart** [comment]
libffi 1.15.0 has been released 🎉
I did not know how to rerun the CI, so I just updated the commit's date without changing its content and force-pushed it.
It's now ready for review.

**Comment by vincentisambart** [comment]
I tried pushing the same change once again with just an updated commit date and the ubuntu truffleruby-head is still failing so that does not seem to be a spurious failure 🤔. I'm wondering if it might be due to some of the ffi 1.15.0 changes.

File: ethon.gemspec
Comment (robnadin) [reply ethon.gemspec:21]: Should this not be `1.15.0`?

Comment (vincentisambart) [reply ethon.gemspec]: Oops, you're totally right. I mixed up 1.5 and 1.15.
Just fixed it.

File: ethon.gemspec
Comment (vincentisambart) [inline ethon.gemspec:21]: Passing callbacks via varargs requires ffi 1.15.0
https://github.com/ffi/ffi/blob/master/CHANGELOG.md#1150--2021-03-05

**Comment by vincentisambart** [comment]
@Kjarrigan Could you have a look?

File: lib/ethon/curls/functions.rb
Comment (matveinazaruk) [inline lib/ethon/curls/functions.rb:20]: I'm still failing on this function. 
Doesn't need any changes here?

```

/home/username/.rvm/gems/ruby-2.4.1/bundler/gems/ethon-90889ce772a9/lib/ethon/easy/operations.rb:44: [BUG] Segmentation fault at 0x007fac9d2d3e20
ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux]

-- Control frame information -----------------------------------------------
c:0028 p:---- s:0154 e:000153 CFUNC  :easy_perform
c:0027 p:0018 s:0149 e:000148 METHOD /home/username/.rvm/gems/ruby-2.4.1/bundler/gems/ethon-90889ce772a9/lib/ethon/easy/operations.rb:44
c:0026 p:0027 s:0145 e:000144 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/typhoeus-1.4.0/lib/typhoeus/request/operations.rb:16
c:0025 p:0041 s:0140 e:000139 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/typhoeus-1.4.0/lib/typhoeus/request/cacheable.rb:18
c:0024 p:0041 s:0135 e:000134 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/typhoeus-1.4.0/lib/typhoeus/request/block_connection.rb:31
c:0023 p:0058 s:0131 e:000130 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/typhoeus-1.4.0/lib/typhoeus/request/stubbable.rb:25
c:0022 p:0027 s:0126 e:000125 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/typhoeus-1.4.0/lib/typhoeus/request/before.rb:26
c:0021 p:0030 s:0122 e:000121 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/typhoeus-1.4.0/lib/typhoeus/request/actions.rb:22
c:0020 p:0053 s:0116 e:000115 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/cocoapods-1.10.1/lib/cocoapods/sources_manager.rb:76
c:0019 p:0020 s:0108 e:000107 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/cocoapods-1.10.1/lib/cocoapods/sources_manager.rb:36
c:0018 p:0020 s:0097 e:000096 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/cocoapods-1.10.1/lib/cocoapods/sources_manager.rb:21
c:0017 p:0013 s:0092 e:000091 BLOCK  /home/username/.rvm/gems/ruby-2.4.1/gems/cocoapods-1.10.1/lib/cocoapods/installer/analyzer.rb:178 [FINISH]
```

Comment (mdimas) [reply lib/ethon/curls/functions.rb:20]: I am still getting an error here as well.

Comment (vincentisambart) [reply lib/ethon/curls/functions.rb:20]: In contrary to the other methods I fixed here, [`curl_easy_perform`](https://curl.se/libcurl/c/curl_easy_perform.html) 's declaration seems correct to me.
I do not have a Linux environment at hand but the tests do run the method that crashed on that stack trace, and the tests did pass on the Ubuntu CI 🤔
Also with or without my change I'm pretty sure ethon is working OK on x86_64 (your stack trace tells me you are running Ruby 2.4.1 on x64_64 in a Linux environment). Without my patch do you have any problem on that environment?

BTW Ruby 2.4 [should not be used anymore](https://www.ruby-lang.org/en/news/2020/04/05/support-of-ruby-2-4-has-ended/) (and  it will be the same for 2.5 very soon), even though I does not look like to be the cause for the crash here.

Comment (matveinazaruk) [reply lib/ethon/curls/functions.rb:20]: Yes, the same errors without your patch.

Comment (mdimas) [reply lib/ethon/curls/functions.rb:20]: I get the same seg faults in the ethon "easy" methods, including  `curl_easy_perform`, with ruby 2.6.5 running on Alpine Linux with musl libc.

Comment (Westacular) [reply lib/ethon/curls/functions.rb:20]: So it appears that the errors on `curl_easy_perform` in Linux are possibly a similar but unrelated issue?

Meanwhile, this PR _does_ fix the segfault on arm64 macOS, which (among other things) is the only remaining blocker to running Cocoapods natively on an M1 Mac. I think it's fair to say this is something a lot of people really want to be able to do.

If you compare the output of the latest failed (truffleruby-head, macOS) test run here with the output from a [recent successful run of it](https://github.com/typhoeus/ethon/pull/179/checks?check_run_id=1947340458), you'll see that the tests did pass for this, but the test harness threw an exception at the end over a thread that didn't converge. This suggests to me that the error was just flakiness in the test harness, and does not reflect any problem with this PR. Previous runs on this MR failed on (truffleruby-head, ubuntu) but passed on (truffleruby-head, macOS), which also suggests that it's flakiness with truffleruby-head and not this PR. This same failure also occurred against master a month ago: https://github.com/typhoeus/ethon/runs/1863459474?check_suite_focus=true

All of which is to say that I don't think there are any real blockers remaining to merging this, and given that it fixes a major crash bug affecting many users... could someone please review and hopefully approve this?

**Review by Kjarrigan — APPROVED** [review]
LGTM although:
- I still can not verify that this solves the Problem on the M1 but @Westacular confirmed that this fixed it so taking his word here
- I'll rollout this MR to some of my production nodes to see if this does not cause any errors on that are not covered by the test-suite
- I just triggered a re-run of the test-suite just to be sure. A failing truffleruby-head is no issue for me. I added this because GitHub provides this env as well but there is no "focus" that it has to always work.

**Comment by Kjarrigan** [comment]
I didn't get any errors after a couple of hours on my nodes so merging this now

**Comment by vincentisambart** [comment]
Thank you!

**Comment by magnusvk** [comment]
@Kjarrigan @vincentisambart thank you for this, this fixes a segfault I was experiencing on an M1 Mac. Any chance you can bake a gem release with this?

**Comment by Kjarrigan** [comment]
Well I'm new to the "maintainer" party and not able to create a release, @i0rek could you please take over and prepare a release?

**Comment by i0rek** [comment]
Thank you everybody for all this work! ❤️ 🚀 https://rubygems.org/gems/ethon/versions/0.13.0.

**Pull request merged**
